crates/driver/src/domain/competition/solution/fee.rs

```diff
@@ -78,21 +78,52 @@ impl Fulfillment {
             Some(FeePolicy::Surplus {
                 factor,
                 max_volume_factor,
+            }) => self.calculate_fee(
+                self.order().sell.amount.0,
+                self.order().buy.amount.0,
+                prices,
+                *factor,
+                *max_volume_factor,
+            ),
+            Some(FeePolicy::PriceImprovement {
+                factor,
+                max_volume_factor,
+                quote,
             }) => {
-                let fee_from_surplus = self.fee_from_surplus(prices, *factor)?;
-                let fee_from_volume = self.fee_from_volume(prices, *max_volume_factor)?;
-                // take the smaller of the two
-                tracing::debug!(uid=?self.order().uid, fee_from_surplus=?fee_from_surplus, fee_from_volume=?fee_from_volume, protocol_fee=?(std::cmp::min(fee_from_surplus, fee_from_volume)), executed=?self.executed(), surplus_fee=?self.surplus_fee(), "calculated protocol fee");
-                Ok(std::cmp::min(fee_from_surplus, fee_from_volume))
+                let sell_amount = quote.sell_amount.max(self.order().sell.amount.0);
+                let buy_amount = quote.buy_amount.min(self.order().buy.amount.0);
+                self.calculate_fee(sell_amount, buy_amount, prices, *factor, *max_volume_factor)
             }
             Some(FeePolicy::Volume { factor }) => self.fee_from_volume(prices, *factor),
             None => Ok(0.into()),
         }
     }
 
-    fn fee_from_surplus(&self, prices: ClearingPrices, factor: f64) -> Result<eth::U256, Error> {
-        let sell_amount = self.order().sell.amount.0;
-        let buy_amount = self.order().buy.amount.0;
+    /// Computes protocol fee compared to the given reference amounts taken from
+    /// the order or a quote.
+    fn calculate_fee(
+        &self,
+        reference_sell_amount: eth::U256,
+        reference_buy_amount: eth::U256,
+        prices: ClearingPrices,
+        factor: f64,
+        max_volume_factor: f64,
+    ) -> Result<eth::U256, Error> {
+        let fee_from_surplus =
+            self.fee_from_surplus(reference_sell_amount, reference_buy_amount, prices, factor)?;
+        let fee_from_volume = self.fee_from_volume(prices, max_volume_factor)?;
+        // take the smaller of the two
+        tracing::debug!(uid=?self.order().uid, fee_from_surplus=?fee_from_surplus, fee_from_volume=?fee_from_volume, protocol_fee=?(std::cmp::min(fee_from_surplus, fee_from_volume)), executed=?self.executed(), surplus_fee=?self.surplus_fee(), "calculated protocol fee");
+        Ok(std::cmp::min(fee_from_surplus, fee_from_volume))
```

Comment:
nit: you can omit `log_name=` from `log_name=?variable_name` if `log_name == variable_name`.
```suggestion
        let protocol_fee = std::cmp::min(fee_from_surplus, fee_from_volume);
        tracing::debug!(uid=?self.order().uid, ?fee_from_surplus, ?fee_from_volume, ?protocol_fee, executed=?self.executed(), surplus_fee=?self.surplus_fee(), "calculated protocol fee");
        Ok(protocol_fee)
```